src/envoy/tcp/alpn_proxy/alpn_proxy.cc

```diff
@@ -0,0 +1,260 @@
+  if (metadata != nullptr) {
+    Envoy::ProtobufWkt::Any metadata_any_value;
+    *metadata_any_value.mutable_type_url() = StructTypeUrl;
+    *metadata_any_value.mutable_value() = metadata->SerializeAsString();
+    std::unique_ptr<::Envoy::Buffer::OwnedImpl> buf =
+        constructProxyHeaderData(metadata_any_value);
+    write_callbacks_->injectWriteDataToFilterChain(*buf, false);
+
+    if (config_->filter_direction_ == FilterDirection::Downstream) {
+      writeMetadata(DownstreamDynamicDataKey, *metadata);
+    } else {
+      writeMetadata(UpstreamDynamicDataKey, *metadata);
+    }
+    config_->stats().metadata_added_.inc();
+  }
+
+  conn_state_ = ReadingInitialHeader;
+}
+
+void AlpnProxyFilter::tryReadInitialProxyHeader(Buffer::Instance& data) {
+  if (conn_state_ != ReadingInitialHeader &&
+      conn_state_ != NeedMoreDataInitialHeader) {
+    return;
+  }
+  const uint32_t initial_header_length = sizeof(AlpnProxyInitialHeader);
+  if (data.length() < initial_header_length) {
+    config_->stats().initial_header_not_found_.inc();
+    // Not enough data to read. Wait for it to come.
+    conn_state_ = NeedMoreDataInitialHeader;
+    return;
+  }
+  std::string initial_header_buf = std::string(
+      static_cast<const char*>(data.linearize(initial_header_length)),
+      initial_header_length);
+  const AlpnProxyInitialHeader* initial_header =
+      reinterpret_cast<const AlpnProxyInitialHeader*>(
+          initial_header_buf.c_str());
+  if (absl::gntohl(initial_header->magic) !=
+      AlpnProxyInitialHeader::magic_number) {
+    config_->stats().initial_header_not_found_.inc();
+    conn_state_ = Invalid;
+    return;
+  }
+  proxy_data_length_ = absl::gntohl(initial_header->data_size);
+  // Drain the initial header length bytes read.
+  data.drain(initial_header_length);
+  conn_state_ = ReadingProxyHeader;
+}
+
+void AlpnProxyFilter::tryReadProxyData(Buffer::Instance& data) {
+  if (conn_state_ != ReadingProxyHeader &&
+      conn_state_ != NeedMoreDataProxyHeader) {
+    return;
+  }
+  if (data.length() < proxy_data_length_) {
+    // Not enough data to read. Wait for it to come.
+    conn_state_ = NeedMoreDataProxyHeader;
+    return;
+  }
+  std::string proxy_data_buf =
+      std::string(static_cast<const char*>(data.linearize(proxy_data_length_)),
+                  proxy_data_length_);
+  Envoy::ProtobufWkt::Any proxy_data;
+  if (!proxy_data.ParseFromString(proxy_data_buf)) {
+    config_->stats().header_not_found_.inc();
+    conn_state_ = Invalid;
+    return;
+  }
+  data.drain(proxy_data_length_);
+
+  Envoy::ProtobufWkt::Struct struct_metadata =
+      Envoy::MessageUtil::anyConvert<Envoy::ProtobufWkt::Struct>(proxy_data);
+  if (config_->filter_direction_ == FilterDirection::Downstream) {
+    writeMetadata(UpstreamDynamicDataKey, struct_metadata);
+  } else {
+    writeMetadata(DownstreamDynamicDataKey, struct_metadata);
+  }
+}
+
+void AlpnProxyFilter::writeMetadata(const std::string key,
```

Comment:
nit: naming suggestion setMetadata so as not to confuse with network writes.

Reply:
Changed.. to be in sync with set.. changed readMetadata to getMetadata too.